### DISCUSSION TITLE
S3 file exists and auth parameters

### DISCUSSION THREAD

**Discussion by ivan-ushakov** [duckdb/duckdb#7571] [Ideas]
For now when you check if S3 file exists, DuckDB does not use S3 parameters because FileOpener is nullptr:
```
bool HTTPFileSystem::FileExists(const string &filename) {
	try {
		auto handle = OpenFile(filename.c_str(), FileFlags::FILE_FLAGS_READ);
		auto &sfh = (HTTPFileHandle &)*handle;
		if (sfh.length == 0) {
			return false;
		}
		return true;
	} catch (const std::exception &e) {
		return false;
	};
}
```

**Comment by juwalter** [comment]
work-around https://github.com/duckdb/duckdb/issues/7572#issuecomment-1554447695

**Comment by quentingodeau** [comment]
Hello @juwalter,

It might be pretty obvious, and I'm missing it..., but how do you get the connection object or the context when you are in FileExists? 
Regards,
Quentin

**Reply by ivan-ushakov** [reply]
Hello,
You have create connection object when you connect to DuckDB instance, usually like this:
```
duckdb::DuckDB db(nullptr); // in memory instance
duckdb::Connection connection(db);
```
After that you can get context and check if file exists:
```
auto &file_opener = duckdb::ClientData::Get(*connection.context).file_opener;
auto is_not_exists = [&fs, &file_opener](const auto &a) {
	try
	{
		auto handle = fs.OpenFile(
			a,
			duckdb::FileFlags::FILE_FLAGS_READ,
			duckdb::FileLockType::NO_LOCK,
			duckdb::FileCompressionType::UNCOMPRESSED,
			file_opener.get());
		LOG_DEBUG() << "File found: " << a;
		return false;
	}
	catch (const std::exception &e)
	{
		LOG_WARNING() << "File does not exist: " << a << ", " << e.what();
		return true;
	};
};
```

**Reply by quentingodeau** [reply]
Hi,

Thx for the reply. But when you are developing an extension (as S3, or in my case the Azure one) you are only providing the FileSystem instantiation, you will not be owner of the connection object (maybe I'm missing something here). From what I see I think that the `OpenerFileSystem` class should as it does others methods, provide the `FileOpener`.
Does it make sense? or once again I'm missing something there?
Regards,
Quentin

**Reply by ivan-ushakov** [reply]
Unfortunately I can not help you with extension development because I don't have such experience. BTW DuckDB has Azure extension.
